### ISSUE TITLE
Code fails if class IDs are not integers starting at 0 with no gaps

### ISSUE THREAD

**Issue by lindsaydbrin** [Dref360/spectral-metric#14]
In `spectral_metric.lib.find_samples()`, starting at line 108:

```
    rng = np.random.RandomState(seed)
    class_samples = []
    class_indices = {}
    indices = np.arange(len(data))
    for k in range(n_class):
        indices_in_cls = rng.permutation(indices[target == k])
        to_take = min(M if M > 1 else int(M * len(indices_in_cls)), len(indices_in_cls))
        class_samples.append(data[indices_in_cls[:to_take]])
        class_indices[k] = indices_in_cls[:to_take]
    return np.array(class_samples), class_indices
```

Line 112 (starting at `indices_in_cls = ...`) shows that the loop is written for `k` to be the class IDs (as in `target`), so it's expecting the class IDs to start with 0 and have no gaps (i.e., the result of `range(n_class)`).  When this is not the case, there will be empty sample arrays, and any classes with indices greater than len(n_class)-1 will not be considered.

As an example, if your set of classes is [9, 13, 15], the loop will iterate over [0, 1, 2] and nothing will be selected for `indices_in_cls`.

As a result, class_samples is an empty array, which makes things fail at sklearn.metrics.pairwise.check_pairwise_arrays() - but also, I don't think this is the desired behaviour.

My first thought was to suggest that the loop be changed to the following:
```
    class_ids = sorted(set(target))
    for class_id in class_ids:
        indices_in_cls = rng.permutation(indices[target == class_id])
        to_take = min(M if M > 1 else int(M * len(indices_in_cls)), len(indices_in_cls))
        class_samples.append(data[indices_in_cls[:to_take]])
        class_indices[class_id] = indices_in_cls[:to_take]
```

But I'm not sure how to think about the fact that this makes `class_indices` a dictionary with keys == actual class IDs, but class_samples an array where each class's samples is appended; in theory you could re-associate class IDs with array indices later, but it feels a bit messy.  You could loop through `for k, class_id in enumerate(class_ids):`, and do `class_indices[k]`, and then re-associate class names later, but I don't know if that makes it better.  

In any case, this association with class IDs and a set of integers from 0 to n_class - 1 has to happen at some point to be able to output things like the s_matrix.  So I'm wondering whether it makes sense to assign classes new IDs early on, keep the rest of the code the same, and then return a dictionary or table later that allows the user to re-associate array positions with class IDs.

**Comment by Dref360** [comment]
I think it would be more correct to have empty arrays. This way you can index the S matrix with the class id, what do you think?

**Comment by lindsaydbrin** [comment]
I think this makes sense!  I think it's a lot safer than any sort of reference table.
